src/lua/protobuf.lua

```diff
@@ -0,0 +1,642 @@
+            'Must be 1'):format(field_def.name, min_key))
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value.'):format(
+            field_def.name, field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_range(field_def, value, range)
+    local min = range ~= nil and range[1] or nil
+    local max = range ~= nil and range[2] or nil
+    -- To avoid incorrect comparison unsigned cdata should not be
+    -- checked for lower limit. If one of the limits is 'nil' this
+    -- function skips the comparison with this limit.
+    if min ~= nil and value < min or max ~= nil and value > max then
+        error(('Input data for %q field is %q and do not fit in %q')
+            :format(field_def.name, value, field_def.type))
+    end
+end
+
+validate_scalar = function(field_def, value)
+    local scalar_def = scalars[field_def.type]
+    local value_type = type(value)
+    assert(scalar_def.accept_type ~= nil)
+    -- Checks type of input according to the allowed types for this field.
+    validate_type(field_def, value, scalar_def.accept_type)
+    -- Checks length of the string if input type assumed to be string.
+    if scalar_def.accept_type == 'string' then
+        validate_length(value)
+    end
+    -- Checks number values for being NaN or inf.
+    if value_type == 'number' and is_nan(value) then
+        error(('Input data for %q field is NaN'):format(field_def.name))
+    end
+    if value_type == 'number' and is_inf(value) then
+        error(('Input data for %q field is inf'):format(field_def.name))
+    end
+    -- Checks values assumed to be integer.
+    if scalar_def.integral_only then
+        check_integer(field_def, value)
+    end
+    -- Checks numeric values to see if they belong to the limits.
+    if scalar_def.limits ~= nil then
+        if value_type == 'cdata' then
+            value_type = ffi.istype(int64_t, value) and 'int64'
+                or 'uint64'
+        end
+        validate_range(field_def, value, scalar_def.limits[value_type])
+    end
+end
+
+-- }}} Validation
+
+
+-- {{{ Scalar type definitions
```

Comment:
What do you think about moving these definitions to line 38?

Reply:
Yes it seems quite logical. Moved this definitions up to the `scalars` declaration.